File: 704_binary_search/binary_search.py
# Py3: https://leetcode.com/problems/binary-search/submissions/1174313614/
import math
from typing import List

def binary_search(array: List[int], start: int, end: int, target: int) -> bool:
    while start <= end:
        mid: int = math.floor(start + (end - start) / 2)
        value: int = array[mid]

        if value == target:
            return True
        elif value < target:
            start = mid + 1
        else:
            end = mid - 1

    return False

numberArray: List[int] = [1,2,3,4,5]
print(binary_search(numberArray, 0, len(numberArray), 2))
print(binary_search(numberArray, 0, len(numberArray), 0))
print(binary_search(numberArray, 0, len(numberArray), 6))

File: 704_binary_search/binary_search.js
// JS: https://leetcode.com/problems/binary-search/submissions/1174319698/
// Deep dive: https://leetcode.com/problems/find-k-th-smallest-pair-distance/solutions/769705/python-clear-explanation-powerful-ultimate-binary-search-template-solved-many-problems/
function binary_search(array, start, end, target) {    
    while(start <= end) {
        let mid = Math.floor(start + (end - start ) / 2);
        let value = array[mid];
        if(value === target) {
            return true;
        } else if (value < target) {
            start = mid + 1;
        } else {
            end = mid - 1;
        }
    }
    return false;
}

let numberArray = [1,2,3,4,5];
console.log(binary_search(numberArray, 0, numberArray.length, 2));
console.log(binary_search(numberArray, 0, numberArray.length, 0));
console.log(binary_search(numberArray, 0, numberArray.length, 6));